core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java

```diff
@@ -3104,23 +3104,106 @@ public static int position(ByteString seek, ByteString s) {
 
   /** SQL {@code POSITION(seek IN string FROM integer)} function. */
   public static int position(String seek, String s, int from) {
-    final int from0 = from - 1; // 0-based
-    if (from0 > s.length() || from0 < 0) {
+    if (from == 0) {
+      throw RESOURCE.fromNotZero().ex();
+    }
+    // Case when from is positive
+    if (from > 0) {
+      final int from0 = from - 1; // 0-based
+      if (from0 >= s.length() || from0 < 0) {
+        return 0;
+      } else {
+        return s.indexOf(seek, from0) + 1;
+      }
+    }
+    // Case when from is negative
+    final int rightIndex = from + s.length(); // negative position to positive index
+    if (rightIndex <= 0) {
       return 0;
     }
-
-    return s.indexOf(seek, from0) + 1;
+    return s.substring(0, rightIndex + 1).lastIndexOf(seek) + 1;
   }
 
   /** SQL {@code POSITION(seek IN string FROM integer)} function for byte
    * strings. */
   public static int position(ByteString seek, ByteString s, int from) {
-    final int from0 = from - 1;
-    if (from0 > s.length() || from0 < 0) {
+    if (from == 0) {
+      throw RESOURCE.fromNotZero().ex();
+    }
+    // Case when from is positive
+    if (from > 0) {
+      final int from0 = from - 1; // 0-based
+      if (from0 >= s.length() || from0 < 0) {
+        return 0;
+      } else {
+        return s.indexOf(seek, from0) + 1;
+      }
+    }
+    // Case when from is negative
+    final int rightIndex = from + s.length();
+    if (rightIndex <= 0) {
       return 0;
     }
+    int lastIndex = 0;
+    while (lastIndex < rightIndex) {
```

Comment:
Is there a method we could add to `ByteString` to allow you to write the same code in this method as for `position(String, String, int)`? (`ByteString` is in Avatica, so we would have to wait a release.)

Reply:
Yes! Java strings have a nice [.lastIndexOf()](https://www.w3schools.com/java/ref_string_lastindexof.asp) function that isn't present in Avatica Bytestrings

Reply:
Ah, good to know. Please log an Avatica bug (similar to https://issues.apache.org/jira/browse/CALCITE-5415) for that. No urgency to fix.

Reply:
https://issues.apache.org/jira/browse/CALCITE-5682?filter=-2
Done! let me know if anything isn't correct there